bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/TableOfContentsFilter.java

```diff
@@ -0,0 +1,305 @@
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.servlets;
+
+import com.adobe.cq.wcm.core.components.internal.models.v1.TableOfContentsImpl;
+import com.adobe.cq.wcm.core.components.models.TableOfContents;
+import com.day.cq.wcm.api.WCMMode;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.servlets.annotations.SlingServletFilter;
+import org.apache.sling.servlets.annotations.SlingServletFilterScope;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.CharArrayWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Intercepts all the HTTP requests made to /editor.html or a html page inside /content/.
+ * Creates a response wrapper - {@link CharResponseWrapper} in which all the servlets/filters called after this filter,
+ * store the response content.
+ * Gets the response content from this wrapper, modifies it and copies it into the original response object.
+ */
+@Component(
+    service = Filter.class,
+    property = {Constants.SERVICE_RANKING + "Integer=999"}
+)
+@SlingServletFilter(
+    scope = {SlingServletFilterScope.REQUEST},
+    pattern = "/content/.*",
+    resourceTypes = "cq:Page",
+    extensions = {"html"},
+    methods = {"GET"}
+)
+public class TableOfContentsFilter implements Filter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TableOfContentsFilter.class);
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+
+    }
+
+    @Override
+    public void destroy() {
+
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+        throws IOException, ServletException {
+
+        CharResponseWrapper responseWrapper = new CharResponseWrapper((HttpServletResponseWrapper) response);
+        chain.doFilter(request, responseWrapper);
+        String originalContent = responseWrapper.toString();
+        Boolean containsTableOfContents = (Boolean) request.getAttribute(TableOfContentsImpl.TOC_REQUEST_ATTR_FLAG);
+        if (responseWrapper.getContentType().contains("text/html") &&
```

Comment:
@jckautzmann we observed NPE in this line during our on-call shift

Reply:
@grubyak thx for your feedback. Could you please create a ticket describing the issue and how it can be reproduced? Thx!